### DISCUSSION TITLE
Logging from a uvm_sequence

### DISCUSSION THREAD

**Discussion by jonpovey** [pyuvm/pyuvm#53] [Q&A] [answered]
I would like to be able to print some debug messages from the body method of a subclass of `uvm_sequence`.
I understand that sequences are transient and not part of the component hierarchy, and don't have the same `self.logger` setup.
Is there a plan for how logging should be done from sequences, sequence items, other such transient things?
It seems like it would be useful if they had a `self.logger` that did something useful, that might be a lazily created logger, or a reference to a shared logger for objects with no hierarchy.

**Comment by raysalemi** [comment]


**Reply by raysalemi** [reply]
I use `uvm_root().logger`

One could also store a logger in `ConfigDB()` from any component.

**Reply by jonpovey** [reply]
Thanks. I am happy with uvm_root().logger for now.

**Reply by pwiecha** [reply]
One downside when using this is that won't get logged to FileHandler set in test. Third option would be: if you instantiate seqs in test, you can pass its logger to them (via std python obj instantiation / init) and add seq name in the log (e.g. decorate) to distinguish between different seqs.

**Reply by raysalemi** [reply]
I wonder if this can be handled with the ConfigDB()

**Comment by jonpovey** [comment, marked as answer]
Just so I can mark an answer:

> use `uvm_root().logger`

**Comment by joaopccarvalho** [comment]
Hello. I am not sure if my question should open a new discussion or not, so I will start here because is seems to me is the closest topic and then I move to another place if needed.

I want to setup the `logging.FileHandler()` where I want to see all log levels. For that I should add:

```python
file_handler = logging.FileHandler(filename='my_log.log')
file_log.setLevel(logging.DEBUG)
self.add_logging_handler(file_log) # 'self' here is a 'uvm_component' child
```

But this only works if before, the logger level is set to a lower level (in this case to the lowest):

```python
self.set_logging_level(logging.DEBUG) # default value is INFO
```

The impact of this will be that I will see all logs on the terminal. But that can be 'fixed' by changing the `_streaming_handler` log level:

```python
self._streaming_handler.setLevel(logging.INFO)
```

My point/question is: shouldn't be better to define the `__default_logging_level = logging.DEBUG` and the `self._streaming_handler.setLevel(logging.INFO)` on `class uvm_report_object(uvm_object)` by default? And so the user can define all new handlers with any level without always having to change the levels of the `logger` and the `_streaming_handler`.

```python
# this wouldn't be needed
self.set_logging_level(logging.DEBUG)

# my handler
file_handler = logging.FileHandler(filename='my_log.log')
file_log.setLevel(logging.DEBUG)
self.add_logging_handler(file_log) # 'self' here is a 'uvm_component' child

# this wouldn't be needed
self._streaming_handler.setLevel(logging.INFO)
```

Not that is a big trouble to do what I show before, but seems less natural. Or maybe adding a comment to how to add a handler will solve everything. Let me know your thought's. Thank you.

**Reply by raysalemi** [reply]
Hi,

I understand why you say that.  The default settings are largely a matter of personal preference,so I decided to mimic the behavior in UVM.
